ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -995,4 +1011,143 @@ void markStartTime() {
       callbackTracker.markOperationStart();
     }
   }
+
+  /**
+   * Injects {@link Account} and {@link Container} for the old put request, which does not carry account or container
+   * header.
+   * @param restRequest The {@link RestRequest} to inject {@link Account} and {@link Container} object.
+   * @throws RestServiceException if either of {@link Account} or {@link Container} object could not be identified.
+   */
+  private void injectAccountAndContainerForOldPutRequest(RestRequest restRequest) throws RestServiceException {
+    Account targetAccount = accountService.getAccountByName(getServiceId(restRequest));
+    if (targetAccount == null) {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
+    }
+    Container targetContainer;
+    if (getIsPrivateSetting(restRequest.getArgs())) {
+      targetContainer = targetAccount.getContainerById(accountService.getContainerIdForLegacyPutPrivateBlob());
+    } else {
+      targetContainer = targetAccount.getContainerById(accountService.getContainerIdForLegacyPutPublicBlob());
+    }
+    if (targetContainer == null) {
+      throw new RestServiceException("Invalid container for putting blob", RestServiceErrorCode.InvalidContainer);
+    }
+    setTargetAccountAndContainerInRestRequest(restRequest, targetAccount, targetContainer);
+  }
+
+  /**
+   * Injects {@link Account} and {@link Container} for the new put request, which carries both account and container
+   * header.
+   * @param restRequest The {@link RestRequest} to inject {@link Account} and {@link Container} object.
+   * @throws RestServiceException if either of {@link Account} or {@link Container} object could not be identified.
+   */
+  private void injectAccountAndContainerForNewPutRequest(RestRequest restRequest) throws RestServiceException {
+    Account targetAccount = accountService.getAccountByName(getTargetAccountNameFromHeader(restRequest));
+    if (targetAccount == null) {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
+    }
+    ensureAccountNameMatchOrThrow(targetAccount, restRequest);
+    Container targetContainer = targetAccount.getContainerByName(getTargetContainerNameFromHeader(restRequest));
+    if (targetContainer == null) {
+      throw new RestServiceException("Invalid container for putting blob", RestServiceErrorCode.InvalidContainer);
+    }
+    ensureContainerNameMatchOrThrow(targetContainer, restRequest);
```

Comment:
Same reason, which is because of the behavior of `InMemoryUnknownAccountService`.